### PR TITLE
[BLOG POST] Blog post about segments, rolling and retention

### PR DESCRIPTION
**Pull request by ppatierno**
### Type of change

* [ ] Typo/minor fix
* [x] New blog post (see the [README](https://github.com/strimzi/strimzi.github.io/#blog-posts) for the process)
* [ ] Other

### CODE REVIEW

**Review by scholzj — COMMENTED** [review]
If you set the data to today or yesterday, we will be able to see the preview -> it will be easier to review it.

File: _posts/2021-12-13-kafka-segment-retention.md
Comment (scholzj) [inline _posts/2021-12-13-kafka-segment-retention.md]: Maybe we can word it without the _everyone knows_ ... it is a bit passive-agressive towards people who might not know ;-)

```suggestion
Apache Kafka behaves as a commit-log when it comes to deal with storing records.
```

**Comment by ppatierno** [comment]
@scholzj done, thanks!

File: _posts/2021-12-13-kafka-segment-retention.md
Comment (scholzj) [inline _posts/2021-12-13-kafka-segment-retention.md]: What about the `.snapshot`? You have it both above and below for one of the segments. But you do not seem to explain it.

File: _posts/2021-12-13-kafka-segment-retention.md
Comment (scholzj) [inline _posts/2021-12-13-kafka-segment-retention.md]: Maybe this is something you can mention early -> that this applies to topics with `delete` policy and not `compact`. When you do it only at the end, it might confuse the users while reading the post.

**Review by scholzj — COMMENTED** [review]
Left some comments after a quick pass ... looks good and helpful.

File: _posts/2021-12-09-kafka-segment-retention.md
Comment (PaulRMellor) [inline _posts/2021-12-09-kafka-segment-retention.md:180]:
```suggestion
Depending on when the last record is appended and a segment is closed, periodic checks for deletion might contribute to missing the 10-minute deadline of the retention period. The closed segment might be deleted on the next check with up to almost 5 minutes delay.
```

File: _posts/2021-12-09-kafka-segment-retention.md
Comment (PaulRMellor) [inline _posts/2021-12-09-kafka-segment-retention.md:182]:
```suggestion
In our example, it might be that the first record in the segment could live up to almost 24 minutes!
```

File: _posts/2021-12-09-kafka-segment-retention.md
Comment (PaulRMellor) [inline _posts/2021-12-09-kafka-segment-retention.md:184]:
```suggestion
Assuming that at some point the clean thread runs and verifies that a closed segment can be deleted. Initially, it adds a `.deleted` extension to the corresponding files, but doesn't actually delete the segment from the file system.
```

File: _posts/2021-12-09-kafka-segment-retention.md
Comment (PaulRMellor) [inline _posts/2021-12-09-kafka-segment-retention.md:185]:
```suggestion
The `log.segment.delete.delay.ms` broker parameter defines when the file will actually be removed from the file system when it's marked as "deleted" (1 minute by default).
```

File: _posts/2021-12-09-kafka-segment-retention.md
Comment (PaulRMellor) [inline _posts/2021-12-09-kafka-segment-retention.md:190]:
```suggestion
Returning to the Canary example again and assuming the delay on deletion, the first record in our segment is still live after 25 minutes!
It is really longer than the 10 minutes expectation, isn't it?
```

File: _posts/2021-12-09-kafka-segment-retention.md
Comment (PaulRMellor) [inline _posts/2021-12-09-kafka-segment-retention.md:192]:
```suggestion
As you can see, it is clear that the retention mechanism doesn't really match the initial expectation.
```

File: _posts/2021-12-09-kafka-segment-retention.md
Comment (PaulRMellor) [inline _posts/2021-12-09-kafka-segment-retention.md:193]:
```suggestion
In reality, a record could live even longer than our 10 minutes depending on the configuration and internal mechanics of the broker.
```

File: _posts/2021-12-09-kafka-segment-retention.md
Comment (PaulRMellor) [inline _posts/2021-12-09-kafka-segment-retention.md:196]:
```suggestion
It is also worth mentioning the impact on the consumer side.
```

File: _posts/2021-12-09-kafka-segment-retention.md
Comment (PaulRMellor) [inline _posts/2021-12-09-kafka-segment-retention.md:197]:
```suggestion
Consumers don't get records from closed segments or deleted segments, even if they are just marked as "deleted" but not actually removed from the file system.
```

File: _posts/2021-12-09-kafka-segment-retention.md
Comment (PaulRMellor) [inline _posts/2021-12-09-kafka-segment-retention.md:199]:
```suggestion
Longer retention won't have a direct impact on the consumers, but more on the disk usage.
```

File: _posts/2021-12-09-kafka-segment-retention.md
Comment (PaulRMellor) [inline _posts/2021-12-09-kafka-segment-retention.md:201]:
```suggestion
The overall retention mechanism works this way when the `log.cleanup.policy` is set to `delete`.
```

File: _posts/2021-12-09-kafka-segment-retention.md
Comment (PaulRMellor) [inline _posts/2021-12-09-kafka-segment-retention.md:202]:
```suggestion
For compacted (`compact`) topics, it is completely different. You can read more on compaction in the [Strimzi documentation](https://strimzi.io/docs/operators/latest/using.html#removing_log_data_with_cleanup_policies).
```

File: _posts/2021-12-09-kafka-segment-retention.md
Comment (PaulRMellor) [inline _posts/2021-12-09-kafka-segment-retention.md:206]:
```suggestion
Understanding how the broker stores partitions and corresponding records on the disk is really important.
```

File: _posts/2021-12-09-kafka-segment-retention.md
Comment (PaulRMellor) [inline _posts/2021-12-09-kafka-segment-retention.md:207]:
```suggestion
Configuration parameters can have a surprisingly big influence on how long your data is retained. 
```

File: _posts/2021-12-09-kafka-segment-retention.md
Comment (PaulRMellor) [inline _posts/2021-12-09-kafka-segment-retention.md:208]:
```suggestion
Learning about these parameters and how you can adjust them gives you a lot more control on how you handle your data.
```

**Review by PaulRMellor — COMMENTED** [review]
Really interesting read. Anything here we want to include in the docs? 
I've made some suggestions. I have a couple of questions where I'm not sure about a sentence.

File: _posts/2021-12-13-kafka-segment-retention.md
Comment (ppatierno) [reply _posts/2021-12-13-kafka-segment-retention.md]: Good advice, thanks!

File: _posts/2021-12-13-kafka-segment-retention.md
Comment (ppatierno) [reply _posts/2021-12-13-kafka-segment-retention.md]: I am not sure about it ... I mean. How could I start the entire paragraph by saying "everything I am going to explain is valid for delete policy and not for the compact one"? Isn't it better just leaving it out until the end as it is now?
Or at least what could be the better way to say it earlier?

File: _posts/2021-12-09-kafka-segment-retention.md
Comment (ppatierno) [reply _posts/2021-12-09-kafka-segment-retention.md:99]: *its* refers to the index file name that follows the same pattern of the log file name (i.e. 00000000000000.index and 0000000000000.log). *it* refers to the index file itself which contains the starting offset of the records indexed by that index.

File: _posts/2021-12-13-kafka-segment-retention.md
Comment (scholzj) [reply _posts/2021-12-13-kafka-segment-retention.md]: Well, your blog post, your call I guess. But while reading it I was wondering all the time about how does it work with compacting and if you are going to explain it. So it felt a bit weird and disappointing at the end. I thought knowing it upfront would make me think about it a bit less. But as I said, it is your call at the end, so do it as you feel.

File: _posts/2021-12-09-kafka-segment-retention.md
Comment (ppatierno) [reply _posts/2021-12-09-kafka-segment-retention.md:130]: No, the meaning is different. My sentence wants to highlight that not always the first record in the segment is the older one.  A producer could send a record with older timestamp even if in the middle of a segment.

File: _posts/2021-12-09-kafka-segment-retention.md
Comment (ppatierno) [reply _posts/2021-12-09-kafka-segment-retention.md:135]: The third condition is the one coming in the next sentence.

File: _posts/2021-12-09-kafka-segment-retention.md
Comment (ppatierno) [reply _posts/2021-12-09-kafka-segment-retention.md:178]: I was referring to the first older record but we could refer to all the records in the segment at this point.

File: _posts/2021-12-13-kafka-segment-retention.md
Comment (ppatierno) [reply _posts/2021-12-13-kafka-segment-retention.md]: @PaulRMellor any suggestion where to put this sentence? Tbh I think it's the right place but having a suggestion could help me to understand if it's better somewhere else.

**Comment by ppatierno** [comment]
@PaulRMellor I  addressed all your suggestions and answered your questions. Would be great to have another pass on them if you have further suggestions. Thanks!

**Review by scholzj — APPROVED** [review]
Love the graphics!!! 😍

File: _posts/2021-12-09-kafka-segment-retention.md
Comment (PaulRMellor) [inline _posts/2021-12-09-kafka-segment-retention.md:16]:
```suggestion
In this blog post, we will dig more into how log segmentation and record retention impacts broker performance. When you know more about how it works, you might want to adjust your log configuration. 

> Another aspect of log configuration is enabling compaction. Using keys to identify messages, Kafka compaction keeps the latest message (with the highest offset) for a specific message key. Earlier messages that have the same key are discarded. You can read more on compaction in the Strimzi documentation for [removing log data with cleanup policies](https://strimzi.io/docs/operators/latest/using.html#removing_log_data_with_cleanup_policies).
```

Comment (PaulRMellor) [reply _posts/2021-12-09-kafka-segment-retention.md:16]: ^^^I've suggested referring to compaction in a note

File: _posts/2021-12-13-kafka-segment-retention.md
Comment (PaulRMellor) [reply _posts/2021-12-13-kafka-segment-retention.md]: I've suggested a compaction note at the start. If you're happy with that, you can remove this.

File: _posts/2021-12-09-kafka-segment-retention.md
Comment (mimaison) [inline _posts/2021-12-09-kafka-segment-retention.md:176]: Since indexes are 10MB by default, this is extreme optimization. I don't think I'd recommend lowering the index size. Indexes are trimmed once they are closed so the gains of doing so are very minimal (with a potential cost of putting a bad value).

File: _posts/2021-12-09-kafka-segment-retention.md
Comment (mimaison) [inline _posts/2021-12-09-kafka-segment-retention.md:178]: What do you mean by `need attention`? What should a user do? Especially as both indexes share the same size setting.

File: _posts/2021-12-09-kafka-segment-retention.md
Comment (mimaison) [inline _posts/2021-12-09-kafka-segment-retention.md:197]: `byes` -> `bytes`

File: _posts/2021-12-09-kafka-segment-retention.md
Comment (mimaison) [inline _posts/2021-12-09-kafka-segment-retention.md:182]: These last 2 paragraphs are not easy to follow. 
What about something like this:
- say the timeindex is the index that can cause roll because it uses a longer keys than other index as described above
- compute how many entries it can contain with the size set
- extrapolate how many records this maps to
- look at example to confirm roll happened before segment was full due to index filling up first

File: _posts/2021-12-09-kafka-segment-retention.md
Comment (ppatierno) [reply _posts/2021-12-09-kafka-segment-retention.md:16]: @PaulRMellor but it's not mentioned or at least not clear that this blog post is valid when policy is "delete". I mean, we are referring to "compact" policy in a note with no context of why we should mention it.

File: _posts/2021-12-09-kafka-segment-retention.md
Comment (ppatierno) [reply _posts/2021-12-09-kafka-segment-retention.md:176]: I agree, but I am not giving any recommendations. I am just highlighting what you should do if you decide to reduce the index file size on your own.

File: _posts/2021-12-09-kafka-segment-retention.md
Comment (ppatierno) [reply _posts/2021-12-09-kafka-segment-retention.md:178]: It's a poor way of writing what I meant for sure :-)
There is no needed intervention from a user pov. It was a kind of introduction to the next sentence ..."Because each ...." so raising the point that you should not focus just on index file but even on timeindex because it can be filled earlier.
So it needs "attention" :-D

File: _posts/2021-12-09-kafka-segment-retention.md
Comment (ppatierno) [reply _posts/2021-12-09-kafka-segment-retention.md:182]: > say the timeindex is the index that can cause roll because it uses a longer keys than other index as described above

I think I addressed it on line 178 with "Because each timeindex entry is 1.5x bigger than an entry in the index (12 bytes versus 8 bytes), it can fill up earlier and cause a new segment to be rolled."

> extrapolate how many records this maps to

Not sure what you mean there is no references to the number of records. I am not mentioning them at all, is that important? I said something like "index file to be filled with one entry every two records." ... maybe this information is not useful at all in this context. We don't want to explain in terms of records but just highlighting that the log segment was rolled before 16k because of the index/timeindex full.

File: _posts/2021-12-09-kafka-segment-retention.md
Comment (PaulRMellor) [reply _posts/2021-12-09-kafka-segment-retention.md:160]: Ah, okay. I've rewritten the suggestion.

File: _posts/2021-12-09-kafka-segment-retention.md
Comment (mimaison) [reply _posts/2021-12-09-kafka-segment-retention.md:176]: To me `you should probably decrease` sounds like a recommendation. @PaulRMellor WDYT?

File: _posts/2021-12-09-kafka-segment-retention.md
Comment (mimaison) [reply _posts/2021-12-09-kafka-segment-retention.md:182]: No worries, that was just a suggestion.

File: _posts/2021-12-09-kafka-segment-retention.md
Comment (PaulRMellor) [inline _posts/2021-12-09-kafka-segment-retention.md:241]:
```suggestion
It is much longer than the 10 minutes expectation, isn't it?
```

**Review by PaulRMellor — APPROVED** [review]
Great :+1:  
I've made a few more suggestion.

File: _posts/2021-12-09-kafka-segment-retention.md
Comment (PaulRMellor) [reply _posts/2021-12-09-kafka-segment-retention.md:16]: I've adjusted the suggestion.

File: _posts/2021-12-09-kafka-segment-retention.md
Comment (PaulRMellor) [reply _posts/2021-12-09-kafka-segment-retention.md:176]: @mimaison - It reads like you would probably do it to be optimal. Maybe better to remove it.

Comment (PaulRMellor) [reply _posts/2021-12-09-kafka-segment-retention.md:176]: Otherwise, maybe:  _Likewise, if you decide to reduce the index file size, it is possible that you might want to decrease the segment size proportionately._

Comment (ppatierno) [reply _posts/2021-12-09-kafka-segment-retention.md:176]: @mimaison I would leave the case but using the sentence proposed by Paul instead of removing it. Wdyt?

**Comment by ppatierno** [comment]
Now that I have all approvals, I rescheduled the publishing data as coming Friday (December 17th) as a good reading for the weekend for people :-)
I will just merge over lunch time CEST to wait US friends wake up.

**Pull request merged**
